### PR TITLE
kraken: Prometheus integration

### PR DESCRIPTION
**Pull request by kinnou02**
This PR adds metrics to kraken than can be exported in the Prometheus format by using this lib: https://github.com/jupp0r/prometheus-cpp

For now, it only measures the number of requests and their duration at the worker level. Others metrics will be added later.
By default, metrics collection and exposition is disabled. The new parameter `GENERAL.metrics_binding` allows us to specify on what address and port the http server must bind by using this form: `127.0.0.1:8080`, this will also enable the collection of the metrics.

After the merge of this PR you will have to update your submodule  recursively:
```
git submodule update --init --recursive
```
We may have to update some of our automated build systems

### CODE REVIEW

File: source/CMakeLists.txt [hunk 1]
```diff
@@ -125,3 +127,7 @@ SET(CMAKE_INSTALL_PREFIX "")
 SET(Rabbitmqc_INCLUDE_DIRS "${CMAKE_SOURCE_DIR}/third_party/librabbitmq-c/librabbitmq/")
 add_subdirectory(third_party/librabbitmq-c)
 add_subdirectory(third_party/SimpleAmqpClient)
+#prometheus-cpp cmake will refuse to build if the CMAKE_INSTALL_PREFIX is empty
+#setting it before will have side effects on how we build packages
+SET(CMAKE_INSTALL_PREFIX "/")
```
Comment (TeXitoi) [inline source/CMakeLists.txt]: Maybe setting something more meaningful as `/usr/local`

File: source/kraken/kraken_zmq.cpp [hunk 2]
```diff
@@ -100,13 +101,16 @@ int main(int argn, char** argv){
         return 1;
     }
 
+    navitia::Metrics metrics(conf.metrics_binding());
+
     threads.create_thread(navitia::MaintenanceWorker(data_manager, conf));
 
+
```
Comment (TeXitoi) [inline source/kraken/kraken_zmq.cpp]: useless empty line

File: source/kraken/kraken_zmq.cpp [hunk 3]
```diff
@@ -100,13 +101,16 @@ int main(int argn, char** argv){
...
     int nb_threads = conf.nb_threads();
     // Launch pool of worker threads
     LOG4CPLUS_INFO(logger, "starting workers threads");
     for(int thread_nbr = 0; thread_nbr < nb_threads; ++thread_nbr) {
-        threads.create_thread(std::bind(&doWork, std::ref(context), std::ref(data_manager), conf));
+        threads.create_thread(std::bind(&doWork, std::ref(context), std::ref(data_manager), conf, std::ref(metrics)));
```
Comment (TeXitoi) [inline source/kraken/kraken_zmq.cpp]: long line

File: source/kraken/kraken_zmq.h [hunk 4]
```diff
@@ -87,7 +90,6 @@ inline void doWork(zmq::context_t& context,
             //on gére le cas du sighup durant un recv
             continue;
         }
-
```
Comment (TeXitoi) [inline source/kraken/kraken_zmq.h]: Any reason to remove this empty line?

File: source/kraken/metrics.cpp [hunk 5]
```diff
@@ -0,0 +1,82 @@
+
+/* Copyright © 2001-2014, Canal TP and/or its affiliates. All rights reserved.
```
Comment (TeXitoi) [inline source/kraken/metrics.cpp]: empty first line + copyright date

File: source/kraken/metrics.cpp [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+    return bucket_boundaries;
+}
+
+
+Metrics::Metrics(const boost::optional<std::string>& endpoint){
+    if(endpoint == boost::none){
+        return;
+    }
+    exposer = std::make_unique<prometheus::Exposer>(*endpoint);
+    registry = std::make_shared<prometheus::Registry>();
```
Comment (TeXitoi) [inline source/kraken/metrics.cpp]: shared finally?

Comment (kinnou02) [reply source/kraken/metrics.cpp]: The `Exposer` take a weak_ptr, so yes shared it goes

**Review by nonifier — APPROVED** [review]

**Review by pbougue — COMMENTED** [review]
Not sure, but you probably have to update scripts and doc elsewhere : https://github.com/CanalTP/navitia/search?utf8=%E2%9C%93&q=git+submodule+update&type=

**Comment by pbougue** [comment]
Also, don't we have to update fabric/jenkins scripts for recursive submodule update?

File: build_navitia.sh [hunk 7]
```diff
@@ -138,7 +138,7 @@ mkdir -p "$run_dir"
 sed -i 's,git\@github.com:\([^/]*\)/\(.*\).git,https://github.com/\1/\2,' .gitmodules
 
 #we need to get the submodules
-git submodule update --init
+git submodule update --init --recursive
```
Comment (xlqian) [inline build_navitia.sh:141]: :+1:

**Review by xlqian — APPROVED** [review]

**Review by benoit-bst — APPROVED** [review]

**Review by TeXitoi — APPROVED** [review]

**Comment by kinnou02** [comment]
jenkins already fetch recursively the submodules.
I already updated navitia-docker-compose, and fabric doesn't use git as it deploy the artifact build by jenkins.

**Review by TeXitoi — APPROVED** [review]

**Review by pbougue — APPROVED** [review]

**Pull request merged**